metrics/storage.go

```diff
@@ -0,0 +1,83 @@
+package metrics
+
+import (
+	"fmt"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// Default service metrics for database operations.
+type StorageMetrics struct {
+	// Name of the runtime (or "consensus") that is being analyzed.
+	runtime string
+
+	// Counts of database operations
+	databaseOperations *prometheus.CounterVec
+
+	// Latencies of database operations.
+	databaseLatencies *prometheus.HistogramVec
+
+	// Cache hit rates for the local cache.
+	localCacheReads *prometheus.CounterVec
+}
+
+type CacheReadStatus string
+
+const (
+	CacheReadStatusHit      CacheReadStatus = "hit"
+	CacheReadStatusMiss     CacheReadStatus = "miss"
+	CacheReadStatusBadValue CacheReadStatus = "bad_value" // Value in cache was not valid (likely because of mismatched types / CBOR encoding).
+	CacheReadStatusError    CacheReadStatus = "error"     // Other internal error reading from cache.
+)
+
+// NewDefaultStorageMetrics creates Prometheus metric instrumentation
+// for basic metrics common to storage accesses.
+func NewDefaultStorageMetrics(runtime string) StorageMetrics {
+	metrics := StorageMetrics{
+		runtime: runtime,
+		databaseOperations: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: fmt.Sprintf("%s_db_operations", runtime),
+				Help: "How many database operations occur, partitioned by operation and status.",
+			},
+			[]string{"database", "operation", "status"}, // Labels.
+		),
+		databaseLatencies: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Name: fmt.Sprintf("%s_db_latencies", runtime),
+				Help: "How long database operations take, partitioned by operation.",
+			},
+			[]string{"database", "operation"}, // Labels.
+		),
+		localCacheReads: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "local_cache_reads",
```

Comment:
Other metrics use names in the pattern of "emerald_local_cache_reads". This is nonidiomatic and makes it hard to aggregate metrics across runtimes. For this new metric, I instead added the runtime as a label (= attribute) of this metric. I'd love to later refactor the other metrics too, bu tthat's more work because we'll also need to update grafana.